sdk/tests/core/test_az_mqtt_connection.c

```diff
@@ -0,0 +1,240 @@
+      ref_conn_rsp++;
+      break;
+    case AZ_MQTT_EVENT_DISCONNECT_RSP:
+      ref_disconn_rsp++;
+      break;
+    case AZ_MQTT_EVENT_PUB_RECV_IND:
+      ref_pub_rcv++;
+      break;
+    case AZ_MQTT_EVENT_PUBACK_RSP:
+      ref_pub_rsp++;
+      break;
+    case AZ_MQTT_EVENT_SUBACK_RSP:
+      break;
+    default:
+      assert_true(false);
+      break;
+  }
+
+  return AZ_OK;
+}
+
+static az_result test_mqtt_connection_callback(az_mqtt_connection* client, az_event event)
+{
+  (void)client;
+  switch (event.type)
+  {
+    case AZ_MQTT_EVENT_CONNECT_RSP:
+      ref_conn_rsp++;
+      break;
+    case AZ_MQTT_EVENT_DISCONNECT_RSP:
+      ref_disconn_rsp++;
+      break;
+    default:
+      assert_true(false);
+      break;
+  }
+
+  return AZ_OK;
+}
+
+static void test_az_mqtt_connection_enabled_init_success(void** state)
+{
+  (void)state;
+
+  test_connection_options = az_mqtt_connection_options_default();
+
+  test_connection_options.hostname = AZ_SPAN_FROM_STR(TEST_HOSTNAME);
+  test_connection_options.client_id_buffer = AZ_SPAN_FROM_STR(TEST_CLIENT_ID);
+  test_connection_options.username_buffer = AZ_SPAN_FROM_STR(TEST_USERNAME);
+  test_connection_options.password_buffer = AZ_SPAN_FROM_STR(TEST_PASSWORD);
+  test_connection_options.port = 8883;
+  test_connection_options.use_username_password = true;
+  test_connection_options.connection_management = true;
+  az_mqtt_x509_client_certificate test_cert = {
+    .cert = AZ_SPAN_FROM_STR("test-cert"),
+    .key = AZ_SPAN_FROM_STR("test-key"),
+  };
+  test_connection_options.client_certificates[0] = test_cert;
+
+  assert_int_equal(az_mqtt_init(&mock_mqtt, &mock_mqtt_options), AZ_OK);
+  assert_int_equal(
+      _az_hfsm_init(
+          &mock_subclient_policy,
+          test_subclient_policy_1_root,
+          test_subclient_policy_get_parent,
+          NULL,
+          NULL),
+      AZ_OK);
+
+  mock_subclient.policy = (az_event_policy*)&mock_subclient_policy;
+
+  assert_int_equal(
+      az_mqtt_connection_init(
+          &test_connection,
+          NULL,
+          &mock_mqtt,
+          test_mqtt_connection_callback,
+          &test_connection_options),
+      AZ_OK);
+
+  assert_int_equal(
+      _az_event_policy_subclients_add_client(
+          &test_connection._internal.subclient_policy, &mock_subclient),
+      AZ_OK);
+}
+
+static void test_az_mqtt_connection_enabled_open_success(void** state)
+{
+  (void)state;
+
+  ref_conn_req = 0;
+  ref_conn_rsp = 0;
+
+  will_return(__wrap_az_platform_clock_msec, 0);
+
+  assert_int_equal(az_mqtt_connection_open(&test_connection), AZ_OK);
+
+  assert_int_equal(ref_conn_req, 1);
+
+  assert_int_equal(
+      az_mqtt_inbound_connack(
+          &mock_mqtt,
+          &(az_mqtt_connack_data){ .connack_reason = 0, .tls_authentication_error = false }),
+      AZ_OK);
+
+  assert_int_equal(ref_conn_rsp, 2);
+}
+
+static void test_az_mqtt_connection_enabled_pub_send_success(void** state)
+{
+  (void)state;
+
+  ref_pub_rsp = 0;
```

Comment:
these tests can't run in parallel, right?

Reply:
We only run separate test processes in parallel.